File: src/entities/vbktx.cpp
// Copyright (c) 2019-2020 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

#include "veriblock/entities/vbktx.hpp"

using namespace altintegration;

void VbkTx::toRaw(WriteStream& stream) const {
  writeNetworkByte(stream, networkOrType);
  sourceAddress.toVbkEncoding(stream);
  sourceAmount.toVbkEncoding(stream);
  stream.writeBE<uint8_t>((uint8_t)outputs.size());
  for (const auto& output : outputs) {
    output.toVbkEncoding(stream);
  }
  writeSingleBEValue(stream, signatureIndex);

  WriteStream pubBytesStream;
  publicationData.toRaw(pubBytesStream);
  writeVarLenValue(stream, pubBytesStream.data());
}

void VbkTx::toVbkEncoding(WriteStream& stream) const {
  WriteStream txStream;
  toRaw(txStream);
  writeVarLenValue(stream, txStream.data());
  writeSingleByteLenValue(stream, signature);
  writeSingleByteLenValue(stream, publicKey);
}

uint256 VbkTx::getHash() const {
  WriteStream stream;
  toRaw(stream);
  return sha256(stream.data());
}

bool altintegration::DeserializeFromRaw(ReadStream& stream,
                                        Slice<const uint8_t> signature,
                                        Slice<const uint8_t> publicKey,
                                        VbkTx& tx,
                                        ValidationState& state) {
  if (!readNetworkByte(stream, TxType::VBK_TX, tx.networkOrType, state)) {
    return state.Invalid("vbktx-network-or-type");
  }

  if (!DeserializeFromVbkEncoding(stream, tx.sourceAddress, state)) {
    return state.Invalid("vbktx-address");
  }
  if (!DeserializeFromVbkEncoding(stream, tx.sourceAmount, state)) {
    return state.Invalid("vbktx-amount");
  }

  uint8_t outputSize = 0;
  if (!stream.readBE<uint8_t>(outputSize, state)) {
    return state.Invalid("vbktx-outputs-size");
  }

  tx.outputs.reserve(outputSize);
  for (size_t i = 0; i < outputSize; i++) {
    Output output;
    if (!DeserializeFromVbkEncoding(stream, output, state)) {
      return state.Invalid("vbktx-output", i);
    }
    tx.outputs.emplace_back(output);
  }

  if (!readSingleBEValue<int64_t>(stream, tx.signatureIndex, state)) {
    return state.Invalid("vbktx-signature-index");
  }

  Slice<const uint8_t> pubBytes;
  if (!readVarLenValue(stream, pubBytes, state, 0, MAX_SIZE_PUBLICATION_DATA)) {
    return state.Invalid("vbktx-publication-bytes");
  }

  if (!DeserializeFromVbkEncoding(pubBytes, tx.publicationData, state)) {
    return state.Invalid("vbktx-publication-data");
  }
  tx.signature = signature.asVector();
  tx.publicKey = publicKey.asVector();

  return true;
}

bool altintegration::DeserializeFromVbkEncoding(ReadStream& stream,
                                                VbkTx& out,
                                                ValidationState& state) {
  Slice<const uint8_t> rawTx;
  if (!readVarLenValue(stream, rawTx, state, 0, MAX_RAWTX_SIZE_VBKTX)) {
    return state.Invalid("vbktx-header");
  }
  Slice<const uint8_t> signature;
  if (!readSingleByteLenValue(
          stream, signature, state, 0, MAX_SIGNATURE_SIZE)) {
    return state.Invalid("vbktx-signature");
  }
  Slice<const uint8_t> publicKey;
  if (!readSingleByteLenValue(stream, publicKey, state, 0, PUBLIC_KEY_SIZE)) {
    return state.Invalid("vbktx-public-key");
  }
  ReadStream txstream(rawTx);
  return DeserializeFromRaw(txstream, signature, publicKey, out, state);
}

File: bindings/go/entities/vbktx.go
// Copyright (c) 2019-2021 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package entities

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"io"

	veriblock "github.com/VeriBlock/alt-integration-cpp/bindings/go"
)

// VbkTx ...
type VbkTx struct {
	NetworkOrType   veriblock.NetworkBytePair
	SourceAddress   Address
	SourceAmount    Coin
	Outputs         []Output
	SignatureIndex  int64
	PublicationData PublicationData
	Signature       []byte
	PublicKey       []byte
}

// GetHash ...
func (v *VbkTx) GetHash() []byte {
	// TODO: Add hash of transaction
	res := make([]byte, 32)
	return res
}

// ToVbkEncoding ...
func (v *VbkTx) ToVbkEncoding(stream io.Writer) error {
	txStream := new(bytes.Buffer)
	if err := v.ToRaw(txStream); err != nil {
		return err
	}
	if err := veriblock.WriteVarLenValue(stream, txStream.Bytes()); err != nil {
		return err
	}
	if err := veriblock.WriteSingleByteLenValue(stream, v.Signature); err != nil {
		return err
	}
	return veriblock.WriteSingleByteLenValue(stream, v.PublicKey)
}

// ToVbkEncodingBytes ...
func (v *VbkTx) ToVbkEncodingBytes() ([]byte, error) {
	var buffer bytes.Buffer
	err := v.ToVbkEncoding(&buffer)
	return buffer.Bytes(), err
}

// ToRaw ...
func (v *VbkTx) ToRaw(stream io.Writer) error {
	if err := v.NetworkOrType.Write(stream); err != nil {
		return err
	}
	if err := v.SourceAddress.ToVbkEncoding(stream); err != nil {
		return err
	}
	if err := v.SourceAmount.ToVbkEncoding(stream); err != nil {
		return err
	}
	if err := binary.Write(stream, binary.BigEndian, uint8(len(v.Outputs))); err != nil {
		return err
	}
	for _, output := range v.Outputs {
		if err := output.ToVbkEncoding(stream); err != nil {
			return err
		}
	}
	if err := veriblock.WriteSingleBEValue(stream, v.SignatureIndex); err != nil {
		return err
	}
	pubBytesStream := new(bytes.Buffer)
	if err := v.PublicationData.ToVbkEncoding(pubBytesStream); err != nil {
		return err
	}
	return veriblock.WriteVarLenValue(stream, pubBytesStream.Bytes())
}

// ToRawBytes ...
func (v *VbkTx) ToRawBytes() ([]byte, error) {
	var buffer bytes.Buffer
	err := v.ToRaw(&buffer)
	return buffer.Bytes(), err
}

// FromVbkEncoding ...
func (v *VbkTx) FromVbkEncoding(stream io.Reader) error {
	rawTx, err := veriblock.ReadVarLenValue(stream, 0, veriblock.MaxRawtxSizeVbktx)
	if err != nil {
		return err
	}
	signature, err := veriblock.ReadSingleByteLenValue(stream, 0, veriblock.MaxSignatureSize)
	if err != nil {
		return err
	}
	publicKey, err := veriblock.ReadSingleByteLenValue(stream, 0, veriblock.PublicKeySize)
	if err != nil {
		return err
	}
	rawTxStream := bytes.NewReader(rawTx)
	return v.FromRaw(rawTxStream, signature, publicKey)
}

// FromVbkEncodingBytes ...
func (v *VbkTx) FromVbkEncodingBytes(data []byte) error {
	buffer := bytes.NewBuffer(data)
	return v.FromVbkEncoding(buffer)
}

// FromRaw ...
func (v *VbkTx) FromRaw(stream io.Reader, signature []byte, publicKey []byte) error {
	if err := v.NetworkOrType.Read(stream, veriblock.TxTypeVbkTx); err != nil {
		return err
	}
	if err := v.SourceAddress.FromVbkEncoding(stream); err != nil {
		return err
	}
	if err := v.SourceAmount.FromVbkEncoding(stream); err != nil {
		return err
	}
	var outputSize uint8
	if err := binary.Read(stream, binary.BigEndian, &outputSize); err != nil {
		return err
	}
	v.Outputs = make([]Output, outputSize)
	for i := 0; i < int(outputSize); i++ {
		if err := v.Outputs[i].FromVbkEncoding(stream); err != nil {
			return err
		}
	}
	if err := veriblock.ReadSingleBEValue(stream, &v.SignatureIndex); err != nil {
		return err
	}
	pubBytes, err := veriblock.ReadVarLenValue(stream, 0, veriblock.MaxSizePublicationData)
	if err != nil {
		return err
	}
	pubBytesStream := bytes.NewReader(pubBytes)
	if err := v.PublicationData.FromVbkEncoding(pubBytesStream); err != nil {
		return err
	}
	v.Signature = signature
	v.PublicKey = publicKey
	return nil
}

// FromRawBytes ...
func (v *VbkTx) FromRawBytes(data []byte, signature []byte, publicKey []byte) error {
	buffer := bytes.NewBuffer(data)
	return v.FromRaw(buffer, signature, publicKey)
}

// ToJSON ...
func (v *VbkTx) ToJSON() (map[string]interface{}, error) {
	outputs := make([]interface{}, len(v.Outputs))
	for i, output := range v.Outputs {
		out, err := output.ToJSON()
		if err != nil {
			return nil, err
		}
		outputs[i] = out
	}
	publicationData, err := v.PublicationData.ToJSON()
	if err != nil {
		return nil, err
	}
	res := map[string]interface{}{
		"hash":            hex.EncodeToString(v.GetHash()),
		"networkByte":     v.NetworkOrType.NetworkByte,
		"type":            v.NetworkOrType.TypeID,
		"sourceAddress":   v.SourceAddress.ToString(),
		"sourceAmount":    v.SourceAmount.Units,
		"outputs":         outputs,
		"signatureIndex":  v.SignatureIndex,
		"publicationData": publicationData,
		"signature":       hex.EncodeToString(v.Signature),
		"publicKey":       hex.EncodeToString(v.PublicKey),
	}
	return res, nil
}
